### PR TITLE
Allow HTTP/2 max concurrent stream configuration

### LINKED ISSUE #8768: gRPC number of concurrent stream block at 250
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.containo.us) and didn't find any.

### What did you do?

I created a simple gRPC client and server. The server is deployed inside our k8s cluster (with k3s) and traefik v2.4.8.
There is only one replica of the server. I got an ingress using h2c protocol and everything works great.

This is my proto, a simple stream from the server to the client:
```proto
service TelemetryService {
  rpc GetTelemetry(TelemetryMeta) returns (stream Telemetry) {}
}
```

When I create a large number of stream from the client by calling GetTelemetry(), the number get blocked to exactly 250.
If I use a NodePort instead of the ingress, everything works as expected.

Moreover, when I close the client, I dont receive the disconnect message in the server. Again, if I use the NodePort, it works.

Some application we have can use up to 5000 concurrent steam.

### What did you see instead?

1) Limit of 250 of concurrent stream
2) No disconnect message in the server when the client is closed

### What version of Traefik are you using?

traefik v2.4.8 with helm chart traefik-9.18.2

### What is your environment & configuration?

```yaml
ports:                                                                                                                                                                                                            
  grpc:                                                                                                                                                                                                  
    port: 50051                                                                                                                                                                                                    
    protocol: TCP                                                                                                                                                                                                  
    expose: true                                                                                                                                                                                                   
    exposedPort: 50051
```

The rest is default from the helm chart.

We use kubernetes with k3s 1.21 deployed on ubuntu 20.04.


### If applicable, please paste the log output in DEBUG level

_No response_

### PR DESCRIPTION
**Pull request by tomMoulard**
### What does this PR do?

This PR allows configuring the max concurrent stream value of HTTP/2 entrypoint.

### Motivation

Fixes #8768

### More

- [X] Added/updated tests
- [X] Added/updated documentation

### CODE REVIEW

File: docs/content/reference/static-configuration/file.toml [hunk 1]
```diff
@@ -30,6 +30,8 @@
       trustedIPs = ["foobar", "foobar"]
     [entryPoints.EntryPoint0.udp]
       timeout = 42
+    [entryPoints.foo.http2]
+      maxConcurrentStreams = 250
```
Comment (rtribotte) [inline docs/content/reference/static-configuration/file.toml:34]:
```suggestion
      maxConcurrentStreams = 42
```

File: docs/content/reference/static-configuration/file.yaml [hunk 2]
```diff
@@ -32,6 +32,8 @@ entryPoints:
       trustedIPs:
       - foobar
       - foobar
+    http2:
+      maxConcurrentStreams: 250
```
Comment (rtribotte) [inline docs/content/reference/static-configuration/file.yaml:36]:
```suggestion
      maxConcurrentStreams: 42
```

File: docs/content/routing/entrypoints.md [hunk 3]
```diff
@@ -127,6 +129,8 @@ They can be defined by using a file (YAML or TOML) or CLI arguments.
     [entryPoints]
       [entryPoints.name]
         address = ":8888" # same as ":8888/tcp"
+        [entryPoints.name.http2]
+          maxConcurrentStreams = 250
```
Comment (rtribotte) [inline docs/content/routing/entrypoints.md:133]:
```suggestion
          maxConcurrentStreams = 42
```

File: docs/content/routing/entrypoints.md [hunk 4]
```diff
@@ -100,6 +100,8 @@ They can be defined by using a file (YAML or TOML) or CLI arguments.
     entryPoints:
       name:
         address: ":8888" # same as ":8888/tcp"
+        http2:
+          maxConcurrentStreams: 250
```
Comment (rtribotte) [inline docs/content/routing/entrypoints.md:104]:
```suggestion
          maxConcurrentStreams: 42
```

File: docs/content/routing/entrypoints.md [hunk 5]
```diff
@@ -148,6 +152,7 @@ They can be defined by using a file (YAML or TOML) or CLI arguments.
     ```bash tab="CLI"
     ## Static configuration
     --entryPoints.name.address=:8888 # same as :8888/tcp
+    --entryPoints.name.http2.maxConcurrentStreams=250
```
Comment (rtribotte) [inline docs/content/routing/entrypoints.md:155]:
```suggestion
    --entryPoints.name.http2.maxConcurrentStreams=42
```

**Review by rtribotte — CHANGES_REQUESTED** [review]

**Review by rtribotte — APPROVED** [review]
LGTM 👌

**Comment by tomMoulard** [comment]
The issue should be fixed, removing the `bot/no-merge` label

**Comment by mpl** [comment]
Summary, explaining why we're delaying on that PR.
we're not seeing any problem at the moment, i.e. the use of x/net seems legitimate for this case, and how we used it seems correct.
However:
- it seems we'll have to keep x/net even more conscientiously updated - at least at every minor Go update, and maybe we have to think more whether that is a cost we are willing to incur
- we're still worried we might be forgetting something, and given we're a reverse-proxy we can't really afford to make mistakes in this area
- and independently, we're realizing that maybe for the h2c case, the http2.Server should have been (even before this PR) initialized with a priority write scheduler, instead of letting it default with a random write scheduler.
TL;DR: we'd like to think about it a bit more, and not rush it for 2.7.

File: docs/content/reference/static-configuration/cli-ref.md [hunk 6]
```diff
@@ -147,8 +147,11 @@ Subject alternative names.
 `--entrypoints.<name>.http.tls.options`:  
 Default TLS options for the routers linked to the entry point.
 
+`--entrypoints.<name>.http2.maxconcurrentstreams`:  
+Specifies the number of concurrent streams that each client may have open at a time. (Default: ```250```)
```
Comment (mpl) [inline docs/content/reference/static-configuration/cli-ref.md:151]:
```suggestion
Specifies the number of concurrent streams per connection that each client is allowed to initiate. (Default: ```250```)
```

File: pkg/config/static/entrypoints.go [hunk 7]
```diff
@@ -59,6 +62,16 @@ type HTTPConfig struct {
 	TLS          *TLSConfig    `description:"Default TLS configuration for the routers linked to the entry point." json:"tls,omitempty" toml:"tls,omitempty" yaml:"tls,omitempty" label:"allowEmpty" file:"allowEmpty" export:"true"`
 }
 
+// HTTP2Config is the HTTP2 configuration of an entry point.
+type HTTP2Config struct {
+	MaxConcurrentStreams int32 `description:"Specifies the number of concurrent streams that each client may have open at a time." json:"maxConcurrentStreams,omitempty" toml:"maxConcurrentStreams,omitempty" yaml:"maxConcurrentStreams,omitempty" export:"true"`
```
Comment (mpl) [inline pkg/config/static/entrypoints.go:67]:
```suggestion
	MaxConcurrentStreams int32 `description:"Specifies the number of concurrent streams per connection that each client is allowed to initiate." json:"maxConcurrentStreams,omitempty" toml:"maxConcurrentStreams,omitempty" yaml:"maxConcurrentStreams,omitempty" export:"true"`
```

File: docs/content/reference/static-configuration/env-ref.md [hunk 8]
```diff
@@ -114,8 +114,11 @@ Trust only forwarded headers from selected IPs.
 `TRAEFIK_ENTRYPOINTS_<NAME>_HTTP`:  
 HTTP configuration.
 
+`TRAEFIK_ENTRYPOINTS_<NAME>_HTTP2_MAXCONCURRENTSTREAMS`:  
+Specifies the number of concurrent streams that each client may have open at a time. (Default: ```250```)
```
Comment (mpl) [inline docs/content/reference/static-configuration/env-ref.md:118]:
```suggestion
Specifies the number of concurrent streams per connection that each client is allowed to initiate. (Default: ```250```)
```

File: docs/content/routing/entrypoints.md [hunk 9]
```diff
@@ -223,6 +228,32 @@ If both TCP and UDP are wanted for the same port, two entryPoints definitions ar
 
     Full details for how to specify `address` can be found in [net.Listen](https://golang.org/pkg/net/#Listen) (and [net.Dial](https://golang.org/pkg/net/#Dial)) of the doc for go.
 
+### HTTP/2
+
+#### `maxConcurrentStreams`
+
+_Optional, Default=250_
+
+`maxConcurrentStreams` defines the number of maximum concurrent streams for HTTP/2.
```
Comment (mpl) [inline docs/content/routing/entrypoints.md:237]:
```suggestion
`maxConcurrentStreams` specifies the number of concurrent streams per connection that each client is allowed to initiate.
```

**Review by kevinpollet — APPROVED** [review]
LGTM

**Review by mpl — APPROVED** [review]

**Comment by MaxThom** [comment]
Hey @tomMoulard, big thanks to you and all your colleagues.
Its working beautifully! No more node ports everywhere :P

**Pull request merged**